File: Classes/Gig.m
//
//  Gig.m
//  FestApp
//

#import "Gig.h"
#import "NSDate+Additions.h"
#import <CoreLocation/CoreLocation.h>

#import "AFNetworking.h"
#import "FestHTTPSessionManager.h"

@interface Gig ()

@end

@implementation Gig

@dynamic timeIntervalString;
@dynamic stageAndTimeIntervalString;
@dynamic duration;

- (instancetype)initFromJSON:(NSDictionary *)json
{
    self = [super init];
    if (self) {
        // http://stackoverflow.com/questions/4999396/how-to-parse-a-date-string-into-an-nsdate-object-in-ios
        NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
        [dateFormatter setDateFormat:@"yyyy-MM-dd'T'HH:mm:ss.SSS'Z'"];

        _begin    = [dateFormatter dateFromString:json[@"start-time"]];
        _end      = [dateFormatter dateFromString:json[@"end-time"]];
        _stage    = json[@"stage"];
        _day      = json[@"day"];

        NSDictionary *artist = json[@"artist"];

		_gigId    = artist[@"id"];
		_gigName  = artist[@"name"];
        _info     = artist[@"info"];

        NSString *wikipediaUrlString = artist[@"wikipedia"];
        if([wikipediaUrlString length] != 0) {
            _wikipediaUrl = [NSURL URLWithString:wikipediaUrlString];
        }

        NSString *imagePath = [NSString stringWithFormat:@"%@.jpg", artist[@"id"]];
        if (imagePath) {
            _imagePath = imagePath;
        }
    }

    return self;
}

- (NSString *)timeIntervalString
{
    if ([self.begin timeIntervalSinceNow] > 18*kOneHour) {
        return [NSString stringWithFormat:@"%@ %@–%@", self.begin.weekdayName, self.begin.hourAndMinuteString, self.end.hourAndMinuteString];
    } else {
        // no need to display weekday name as a part of a current-day gig:
        return [NSString stringWithFormat:@"%@–%@", self.begin.hourAndMinuteString, self.end.hourAndMinuteString];
    }
}

- (NSString *)stageAndTimeIntervalString
{
    NSString *end = [self.end.hourAndMinuteString isEqualToString:@"23:59"] ? @"00:00" : self.end.hourAndMinuteString;
    return [NSString stringWithFormat:@"%@ %@–%@ %@", self.begin.weekdayName, self.begin.hourAndMinuteString, end, self.stage];
}

- (NSTimeInterval)duration
{
	return [self.end timeIntervalSinceDate:self.begin];
}

- (NSString *)debugDescription
{
    return self.gigName;
}

@end

File: Classes/Gig.swift
//
//  Gig.swift
//  FestApp
//
//  Created by Oleg Grenrus on 29/10/14.
//  Copyright (c) 2014 Futurice Oy. All rights reserved.
//

import Foundation

class Gig: NSObject {
    let begin: NSDate
    let end: NSDate
    let stage: String
    let day: String

    let gigId: String
    let gigName: String
    let info: String

    let imagePath: String?
    let wikipediaUrl: NSURL?

    init(dictionary: NSDictionary) {
        begin = (dictionary["start-time"] as? String)?.asIsoDate() ?? swiftDistantPast
        end   = (dictionary["end-time"] as? String)?.asIsoDate() ?? swiftDistantPast

        stage = dictionary["stage"] as? String ?? ""
        day   = dictionary["day"] as? String ?? ""

        let artist: NSDictionary? = dictionary["artist"] as? NSDictionary

        gigId = artist?["id"] as? String ?? ""
        gigName = artist?["name"] as? String ?? ""
        info = artist?["info"] as? String ?? ""

        let wikipediaString: Optional<String> = artist?["wikipedia"] as? String

        // extension magic!
        let str: NSURL? = wikipediaString
            .filter({ str in !str.isEmpty })
            .flatmap({ str in NSURL(string: str, relativeToURL: nil) })

        imagePath = (artist?["id"] as? String).map({ artistId in artistId + ".jpg" })
    }

    func timeIntervalString() -> String {
        return ""
    }

    func stageAndTimeIntervalString() -> String {
        // String interpolation
        return "\(day), \(dateHourAndMinuteString(begin))-\(dateHourAndMinuteString(end)) \(stage)"
    }

    func duration() -> NSTimeInterval {
        return end.timeIntervalSinceDate(begin)
    }
}
